pkg/build/controller/build/build_controller.go

```diff
@@ -0,0 +1,850 @@
+	case buildutil.IsBuildComplete(build):
+		update, err = bc.handleCompletedBuild(build, pod, podErr)
+	}
+	if update != nil && !update.isEmpty() {
+		bc.updateBuild(build, update, pod)
+	}
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// shouldIgnore returns true if a build should be ignored by the controller.
+// These include pipeline builds as well as builds that are in a terminal state.
+// However if the build is either complete or failed and its completion timestamp
+// has not been set, then it returns false so that the build's completion timestamp
+// gets updated.
+func shouldIgnore(build *buildapi.Build) bool {
+	// If pipeline build, do nothing.
+	// These builds are processed/updated/etc by the jenkins sync plugin
+	if build.Spec.Strategy.JenkinsPipelineStrategy != nil {
+		glog.V(4).Infof("Ignoring build %s with jenkins pipeline strategy", buildDesc(build))
+		return true
+	}
+
+	// If a build is in a terminal state, ignore it; unless it is in a succeeded or failed
+	// state and its completion time is not set, then we should at least attempt to set its
+	// completion time if possible.
+	if buildutil.IsBuildComplete(build) {
+		switch build.Status.Phase {
+		case buildapi.BuildPhaseComplete,
+			buildapi.BuildPhaseFailed:
+			if build.Status.CompletionTimestamp == nil {
+				return false
+			}
+		}
+		glog.V(4).Infof("Ignoring build %s in completed state", buildDesc(build))
+		return true
+	}
+
+	return false
+}
+
+// shouldCancel returns true if a build is active and its cancellation flag is set
+func shouldCancel(build *buildapi.Build) bool {
+	return !buildutil.IsBuildComplete(build) && build.Status.Cancelled
+}
+
+// cancelBuild deletes a build pod and returns an update to mark the build as cancelled
+func (bc *BuildController) cancelBuild(build *buildapi.Build) (*buildUpdate, error) {
+	glog.V(4).Infof("Cancelling build %s", buildDesc(build))
+
+	podName := buildapi.GetBuildPodName(build)
+	err := bc.podClient.Pods(build.Namespace).Delete(podName, &metav1.DeleteOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return nil, fmt.Errorf("could not delete build pod %s/%s to cancel build %s: %v", build.Namespace, podName, buildDesc(build), err)
+	}
+
+	update := &buildUpdate{}
+	update.setPhase(buildapi.BuildPhaseCancelled)
+	update.setReason(buildapi.StatusReasonCancelledBuild)
+	update.setMessage(buildapi.StatusMessageCancelledBuild)
+
+	return update, nil
+}
+
+// handleNewBuild will check whether policy allows running the new build and if so, creates a pod
+// for the build and returns an update to move it to the Pending phase
+func (bc *BuildController) handleNewBuild(build *buildapi.Build, pod *kapi.Pod, podErr error) (*buildUpdate, error) {
+	// If a pod was found, and it was created after the build was created, it
+	// means that the build is active and its status should be updated
+	if podErr == nil && pod != nil {
+		if !pod.CreationTimestamp.Before(build.CreationTimestamp) {
```

Comment:
I wonder if the error handling in createBuildPod can be brought up to here so that the existing pod case is handled in one place, whether older than the build object or not.